### DISCUSSION TITLE
createTileCache options

### DISCUSSION THREAD

**Discussion by moving-bits** [mapsforge/mapsforge#1545] [Q&A] [answered]
While cross-checking the updates samples for map rotation (thanks for posting https://github.com/cgeo/cgeo/issues/15743#issuecomment-2360483349) I've stumbled upon `getScreenRatio()` (set to `2f` to increase tile cache size) and `createTileCache()` options. 

We call `createTileCache()` in `onViewCreated()` of the map fragment, so before any theme got loaded. Would it be safe to make this cache persistent even when changing the theme later on - does Mapsforge handle this internally? From a quick test it seems to be ok, but I'd rather ask...

**Comment by devemux86** [comment, marked as answer]
@moving-bits 

- We need a larger tile cache because of the map rotation, via API or gesture.
- We **may** need a larger overdraw factor if we change the map view center of rotation.

If you ask if the tile cache can be persistent or not, it can be set during its creation.
You can clear the cache folder when changing the theme or use different cache folder per theme.

This depends on how the app wants to handle the tile caches. Mapsforge just offers the options.

**Reply by moving-bits** [reply]
Thanks for the fast reply.

> * We need a larger tile cache because of the map rotation, via API or gesture.

I've set it to `2f` locally already (like in the updated samples), will push a PR for c:geo for this.

> * We **may** need a larger overdraw factor if we change the map view center of rotation.

We have a "driving mode" (with automated map rotation) which sets map center to 25% of screen's height from the bottom. Would this require an overview factor other than the default `mMapView.getModel().frameBufferModel.getOverdrawFactor()`?

> If you ask if the tile cache can be persistent or not, it can be set during its creation. You can clear the cache folder when changing the theme or use different cache folder per theme.

So theme change is not handled by the cache internally? Is there some documentation on what factors require a cache purge?

I guess I need to purge the cache on map source change or theme change. But does it have to be purged on theme options change as well? (And what would be a good place for doing so? Currently we purge on each `onStart` to make sure changed theme options get applied, but there's a less ressource-intensive way probably.)

**Reply by devemux86** [reply]
@moving-bits 

> Would this require an overview factor other than the default mMapView.getModel().frameBufferModel.getOverdrawFactor()?

`RotationMapViewer` uses such map view center, so can check its code.

About the tile cache the simpler is to clear it when changing theme or layers.
Otherwise keep it.

Such actions must be done by the apps, they know their themes, their layers and what to do with them.

**Reply by moving-bits** [reply]
> `RotationMapViewer` uses such map view center, so can check its code.

If I read the sample app correctly, it creates tile cache first (with the then existing overdraw factor), and later on set a new overdraw factor (which does not get taken into account for tile cache calculation as that is only done once on activity's `onCreate()`).

This would fit into c:geo's current tile cache creation scheme, as it's created early in the process as well. Only re-setting the overdraw factor when entering driving mode is missing yet - I will add that.

Other than those two (factor `2f` for tile cache size and re-setting overdraw factor) I will leave tile cache handling "as is" for now. Thanks for your help (and for all the work that has gone into updating Mapsforge, very welcomed! - just read the "Cruiser 5 beta announcement thread today...).

**Reply by devemux86** [reply]
@moving-bits thanks for reporting this.

This is probably wrong, the overdraw factor better be set once at start so it can be used everywhere.
So we should move this statement before the cache creation.

**Reply by moving-bits** [reply]
But overdraw factor seems to be depending on the map view center being set. At startup, this is `0.5f`, and that's the base used for calculating the tile cache size.
But the center can be changed later on (and with it the overdraw factor) - so tile cache size needs to be adjusted as well.

Or we need some different algorithm for calculating the value of "overdraw" parameter at tile cache creation time, which takes the possibility of map center = `0.75f` into account. (So it cannot be made depending on `mMapView.getModel().frameBufferModel.getOverdrawFactor()` directly, which has the wrong value at that point in time.)

**Reply by devemux86** [reply]
@moving-bits 

The overdraw factor is used in tile cache calculations, so it is better to set it before.
You can still change it during app usage, but the cache better use the larger value.

(I updated the sample)

**Comment by devemux86** [comment]
@moving-bits 

Remember that the tile cache stores only the map tiles underneath.

If you use the label layer, this is separate from the cache, it remains free to render and change during rotation.

**Reply by moving-bits** [reply]
Yes, we use `LabelLayer` in our new `UnifiedMap` Mapsforge implementation for offline-based maps.
(The "old" Mapsforge-based map - internally called `NewMap` - doesn't support rotation, therefore also no `LabelLayer`.)
